### PR TITLE
Feature/extend create product window

### LINKED ISSUE #31: Extend create product window
Add missing *optional fields from the product to the window to allow users to enter those values

- add validation based on the field type(e.g. if it is a number - allow only numbers) but still keep those fields optional

### PR DESCRIPTION
**Pull request by ghost**
Description
Create product form extended - new optional inputs added - location, quantityAvailable, price.

Screen recording and before screenshot
https://drive.google.com/drive/folders/1ETAzap_Bp-0wRh43Go2wdFTpAdkGcrJA?usp=sharing

Issue
Fixes: https://github.com/vnazarchukGD/angular-time-killer/issues/31

### CODE REVIEW

File: src/app/shared/modules/create-product-modal/create-product-modal.component.html [hunk 1]
```diff
@@ -31,6 +31,46 @@ <h1 mat-dialog-title>Create new product</h1>
         {{ getErrorMessage("description") }}
       </mat-error>
     </mat-form-field>
+
+    <div formArrayName="counts">
+      <div
+        class="form-array"
+        *ngFor="let count of counts().controls; let i = index"
+        [formGroupName]="i"
+      >
```
Comment (ooow) [inline src/app/shared/modules/create-product-modal/create-product-modal.component.html:40]: Please update formatting here from:

```
     <div
        class="form-array"
        *ngFor="let count of counts().controls; let i = index"
        [formGroupName]="i"
      >   
```

to 


```
     <div *ngFor="let count of counts().controls; let i = index"
          class="form-array"
          [formGroupName]="i">
```

Here and everywhere else(similar cases)

File: src/app/shared/modules/create-product-modal/create-product-modal.component.html [hunk 2]
```diff
@@ -31,6 +31,46 @@ <h1 mat-dialog-title>Create new product</h1>
...
+        <div>
+          <mat-form-field appearance="fill">
+            <mat-label>Location</mat-label>
+            <input matInput type="text" formControlName="location" />
```
Comment (ooow) [inline src/app/shared/modules/create-product-modal/create-product-modal.component.html:44]: Don't leave a ` ` (space) before closing tag here and everywhere else.
`location" />` -> `location"/>`

File: src/app/shared/modules/create-product-modal/create-product-modal.component.html [hunk 3]
```diff
@@ -31,6 +31,46 @@ <h1 mat-dialog-title>Create new product</h1>
...
+            </mat-form-field>
+            <mat-form-field appearance="fill">
+              <mat-label>Price</mat-label>
+              <input matInput type="number" formControlName="price" />
+            </mat-form-field>
+          </div>
+        </div>
+        <button (click)="removeCountsRecord(i)" mat-icon-button
+          ><mat-icon>remove_circle_outline</mat-icon></button
+        >
```
Comment (ooow) [inline src/app/shared/modules/create-product-modal/create-product-modal.component.html:63]: Please move <button to the next line and close it right there.

File: src/app/shared/modules/create-product-modal/create-product-modal.component.scss [hunk 4]
```diff
@@ -16,6 +16,29 @@
   }
 }
 
+.form-array {
+  display: flex;
+  justify-content: space-between;
+  align-items: center;
+  width: 100%;
```
Comment (ooow) [inline src/app/shared/modules/create-product-modal/create-product-modal.component.scss:23]: Please keep alphabetical order here and everywhere else(in similar cases)

File: src/app/shared/modules/create-product-modal/create-product-modal.component.spec.ts [hunk 5]
```diff
@@ -16,7 +21,7 @@ import { CreateProductModalModule } from 'src/app/shared/modules/create-product-
 import * as imageUrlValidator from 'src/app/shared/modules/create-product-modal/validators/image-url.validator';
 import Spy = jasmine.Spy;
 
-describe('CreateProductModalComponent', () => {
+fdescribe('CreateProductModalComponent', () => {
```
Comment (ooow) [inline src/app/shared/modules/create-product-modal/create-product-modal.component.spec.ts:24]: Please remove *f* prefix

File: src/app/shared/modules/create-product-modal/create-product-modal.component.spec.ts [hunk 6]
```diff
@@ -74,8 +92,13 @@ describe('CreateProductModalComponent', () => {
     expect(await harness.nameFormField()).toBeTruthy();
     expect(await harness.pictureUrlFormField()).toBeTruthy();
     expect(await harness.descriptionFormField()).toBeTruthy();
+    expect(await harness.locationFormField()).toBeTruthy();
+    expect(await harness.quantityAvailableFormField).toBeTruthy();
```
Comment (ooow) [inline src/app/shared/modules/create-product-modal/create-product-modal.component.spec.ts:96]: Did you forget to call the function? here and bellow

File: src/app/shared/modules/create-product-modal/create-product-modal.component.spec.ts [hunk 7]
```diff
@@ -313,5 +337,33 @@ describe('CreateProductModalComponent', () => {
         ).toBe('test test');
       });
     });
+
+    describe('Create new Location Btn', () => {
```
Comment (ooow) [inline src/app/shared/modules/create-product-modal/create-product-modal.component.spec.ts:341]: Only root describe suppose to start with Upper case, and all others depend on it. E.g.

```
describe('ExampleCalss', () => {
  describe('with some property', () => {
     it('does something', () => {
        ...
     });
  });

  describe('without some property', () => {
     it('does something', () => {
        ...
     });
  });
});
```

Turns to >>  2 cases:
1. ***ExampleCalss with some property does something***
2. ***ExampleCalss without some property does something***

The main goal is to make at the end a sentence that makes sense.

File: src/app/shared/modules/create-product-modal/create-product-modal.component.spec.ts [hunk 8]
```diff
@@ -313,5 +337,33 @@ describe('CreateProductModalComponent', () => {
...
+      it('should add new  record to counts property - form', fakeAsync(() => {
```
Comment (ooow) [inline src/app/shared/modules/create-product-modal/create-product-modal.component.spec.ts:342]: Why do you need *fakeAsync* here?

File: src/app/shared/modules/create-product-modal/create-product-modal.component.spec.ts [hunk 9]
```diff
@@ -313,5 +337,33 @@ describe('CreateProductModalComponent', () => {
...
+        tick();
+        fixture.detectChanges();
+
+        expect(
+          component.dialogRef.componentInstance.addCountsRecord
+        ).toHaveBeenCalled();
+      }));
+    });
+
+    describe('Remove location Btn', () => {
```
Comment (ooow) [inline src/app/shared/modules/create-product-modal/create-product-modal.component.spec.ts:355]: This description splitting does not make big sense here, better remove those and just leave to ***it*** cases for simplicity.

File: src/app/shared/modules/create-product-modal/create-product-modal.component.ts [hunk 10]
```diff
@@ -39,11 +40,38 @@ export class CreateProductModalComponent implements OnDestroy {
           Validators.maxLength(500),
           Validators.pattern(TEXT_REGEX_PATTERN)
         ]
-      ]
+      ],
+      counts: this.fb.array([
+        this.fb.group({
+          location: [''],
+          quantityAvailable: ['', Validators.pattern(NUMERIC_REGEX_PATTERN)],
```
Comment (ooow) [inline src/app/shared/modules/create-product-modal/create-product-modal.component.ts:47]: Please don't use regexps when they are not necessary. here you can just check that typeof is "number" or use isNaN function


Apply to other places as well.

File: src/app/shared/modules/create-product-modal/create-product-modal.component.ts [hunk 11]
```diff
@@ -39,11 +40,38 @@ export class CreateProductModalComponent implements OnDestroy {
...
+          price: ['', Validators.pattern(NUMERIC_REGEX_PATTERN)]
+        })
+      ])
     },
     { updateOn: 'blur' }
   );
 
+  counts(): FormArray {
+    return this.form.get('counts') as FormArray;
```
Comment (ooow) [inline src/app/shared/modules/create-product-modal/create-product-modal.component.ts:56]: In general, it is not a good style to write single-line methods. Please use directly *this.form.get('counts')* and do not use ***as*** casting - it is forbidden and considered as bad practice.

File: src/app/shared/modules/create-product-modal/create-product-modal.component.ts [hunk 12]
```diff
@@ -39,11 +40,38 @@ export class CreateProductModalComponent implements OnDestroy {
...
+  newCountsRecord(): FormGroup {
+    return this.fb.group({
+      location: [''],
+      quantityAvailable: ['', Validators.pattern(NUMERIC_REGEX_PATTERN)],
+      price: ['', Validators.pattern(NUMERIC_REGEX_PATTERN)]
+    });
+  }
+
+  addCountsRecord() {
+    this.counts().push(this.newCountsRecord());
```
Comment (ooow) [inline src/app/shared/modules/create-product-modal/create-product-modal.component.ts:68]: ditto(about single line methods)

File: src/app/shared/modules/create-product-modal/create-product-modal.component.ts [hunk 13]
```diff
@@ -39,11 +40,38 @@ export class CreateProductModalComponent implements OnDestroy {
...
+  }
+
+  removeCountsRecord(index: number) {
+    this.counts().removeAt(index);
```
Comment (ooow) [inline src/app/shared/modules/create-product-modal/create-product-modal.component.ts:72]: ditto(about single line methods)

File: src/app/shared/modules/create-product-modal/create-product-modal.component.ts [hunk 14]
```diff
@@ -130,8 +158,9 @@ export class CreateProductModalComponent implements OnDestroy {
     return (
       this.form.invalid ||
       !this.formValues ||
-      Object.values(this.formValues).filter(Boolean).length !==
-        Object.values(this.form.value).length
+      Object.values(this.form.controls).some((field) =>
```
Comment (ooow) [inline src/app/shared/modules/create-product-modal/create-product-modal.component.ts:161]: The button should be disabled only when *this.form.valid* is false - if this is not working - then probably form validators are not set correctly according to task requirements. Please check and update.

**Review by ooow — CHANGES_REQUESTED** [review]
Looks good, but needs tiny imporvments

File: src/app/shared/modules/create-product-modal/create-product-modal.component.html (hunk 2 above)
Comment (mmtasarz) [reply src/app/shared/modules/create-product-modal/create-product-modal.component.html:44]: I checked all formatting rules, but I can't find where I can change or set it, also I checked some forums etc. but I didn't find the solution

**Comment by mmtasarz** [comment]
I resolved issues from comments, but i have a problem with 2 tests, I have no idea why they don't pass (CreateProductModalComponent > Create Button > should close dialog and emit valid form values on create btn click and Header > Create Product Modal > should call product service stored-product method on create product with expected params)

**Pull request closed without merging**